src/lib.rs

```diff
@@ -515,228 +509,234 @@ impl TimeWindow {
     }
 
     #[must_use]
-    pub fn duration(&self) -> Duration {
+    pub fn length(&self) -> Duration {
         self.end - self.start
     }
 
+    /// Creates a new `TimeWindow` with `start` set to `new_start`. If
+    /// `new_start` is greater than or equal to `end` the start will be set
+    /// equal to `end`.
     #[must_use]
-    pub fn with_new_start(self, start: Time) -> TimeWindow {
-        TimeWindow::new(start, self.end)
+    pub fn with_start(&self, new_start: Time) -> Self {
+        Self::new(new_start.min(self.end), self.end)
     }
 
+    /// Creates a new `TimeWindow` with `end` set to `new_end`. If `new_end` is
+    /// smaller or equal to `start`, the `end` will be set to `start.`
     #[must_use]
-    pub fn with_new_end(self, end: Time) -> TimeWindow {
-        TimeWindow::new(self.start, end)
+    pub fn with_end(&self, new_end: Time) -> Self {
+        Self::new(self.start, new_end.max(self.start))
     }
 
-    /// Extends time window start to the given value. Is a No-Op when given
-    /// value isn't earlier than current time window start.
-    /// Returns by which duration the start was moved.
-    /// # Examples
+    /// Creates a new `TimeWindow` with the `start` preponed to the given value.
+    /// If `new_start` isn't earlier than the current time window start, a copy
+    /// of `self` is returned.
     ///
+    /// # Examples
     /// ```
-    /// # use tinytime::Duration;
-    /// # use tinytime::Time;
-    /// # use tinytime::TimeWindow;
-    /// let mut x = TimeWindow::from_seconds(4, 5);
-    /// assert_eq!(Some(Duration::seconds(1)), x.extend_start(Time::seconds(3)));
-    /// assert_eq!(Time::seconds(3), x.start());
-    /// assert_eq!(None, x.extend_start(Time::seconds(6)));
-    /// assert_eq!(Time::seconds(3), x.start());
+    /// # use tinytime::*;
+    /// let x = TimeWindow::from_seconds(4, 5);
+    /// assert_eq!(
+    ///     TimeWindow::from_seconds(3, 5),
+    ///     x.prepone_start_to(Time::seconds(3))
+    /// );
+    /// assert_eq!(
+    ///     TimeWindow::from_seconds(4, 5),
+    ///     x.prepone_start_to(Time::seconds(6))
+    /// );
     /// ```
     #[must_use]
-    pub fn extend_start(&mut self, new_start: Time) -> Option<Duration> {
-        (new_start < self.start).then(|| {
-            let diff = self.start - new_start;
-            *self = self.with_new_start(new_start);
-            diff
-        })
+    pub fn prepone_start_to(&self, new_start: Time) -> Self {
+        (new_start < self.start)
+            .then(|| self.with_start(new_start))
+            .unwrap_or(*self)
```

Comment:
I think `self.with_start(self.start.min(new_start))` would be a bit easier to follow (same for the other `prepone_*_to` and `postpone_*_to` methods).

Reply:
I'm not sure we need these methods at all... They feel too specific. Maybe let's move them into application instead?

Reply:
I refactored the code, indeed much simpler.

@urmaul I think/hope that with the new consistent names having these (arguably simple) methods is a net benefit because it makes the code easy to read and the library easy to use. Let's evaluate that in practice and if not we can do another refactoring and maybe remove stuff.